stock_available_mrp/models/product_product.py

```diff
@@ -0,0 +1,126 @@
+        bom_obj = self.env['mrp.bom']
+        bom = bom_obj._bom_find(product=product)
+        if not bom:
+            return 0.0
+
+        # Need by product (same product can be in many BOM lines/levels)
+        try:
+            component_needs = self._get_components_needs(product, bom)
+        except AccessError:
+            # If user doesn't have access to BOM
+            # he can't see potential_qty
+            component_needs = None
+        if not component_needs:
+            # The BoM has no line we can use
+            product.potential_qty = 0.0
+
+        else:
+            # Find the lowest quantity we can make with the stock at hand
+            components_potential_qty = min(
+                [self._get_component_qty(component) // need
+                 for component, need in component_needs.items()]
+            )
+
+            # Compute with bom quantity
+            bom_qty = bom.product_uom_id._compute_quantity(
+                bom.product_qty,
+                bom.product_tmpl_id.uom_id
+            )
+            return bom_qty * components_potential_qty
+
+    def _get_component_qty(self, component):
+        """ Return the component qty to use based en company settings.
+
+        :type component: product_product
+        :rtype: float
+        """
+        icp = self.env['ir.config_parameter'].sudo()
+        stock_available_mrp_based_on = icp.get_param(
+            'stock_available_mrp_based_on', 'qty_available'
+        )
+
+        return component[stock_available_mrp_based_on]
+
+    def _get_components_needs(self, product, bom):
+        """ Return the needed qty of each compoments in the *bom* of *product*.
+
+        :type product: product_product
+        :type bom: mrp_bom
+        :rtype: collections.Counter
+        """
+        needs = Counter()
+        for bom_component in bom.explode(product, 1.0)[1]:
+            component = bom_component[0]['product_id']
+            try:
+                line_uom = bom_component[0].product_uom_id
+                component_qty = line_uom._compute_quantity(
+                    bom_component[1]['original_qty'],
+                    component.uom_id
+                )
+            except UserError:
+                # This occurs, when we have components
+                # which are in a different cateogry.
+                continue
+            component_qty = component_qty * bom_component[1]['qty']
+            needs += Counter(
+                {component: component_qty}
+            )
+
+        return needs
+
+    def _get_component_ids(self):
+        """ Compute component_ids by getting all the components for
+        this product.
+        """
+        bom_obj = self.env['mrp.bom']
+
+        bom_id = bom_obj._bom_find(product_id=self.id)
+        if bom_id:
+            bom = bom_obj.browse(bom_id)
+            for bom_component in bom_obj._bom_explode(bom, self, 1.0)[0]:
+                self.component_ids |= self.browse(bom_component['product_id'])
```

Comment:
should be @api.multi + loop and collect all the product ids and browse only once at the end.

Reply:
I think we can drop this compute function since it is currently never called since component_ids is never accessed directly. component_ids was used for an api.depends so we would update the quantity if it changed on one of the components. But since the potential_qty isn't stored we would recalculate it anyways if accessed. Do you agree or am i missing something? /cc @grindtildeath

Reply:
@tschanzt Seems to make sense although I fear performance could be an issue not storing `potential_qty`...
Let us have a look together on monday :wink:

Reply:
@tschanzt Allright let's remove this function and `_compute_available_quantities` and we'll be good ! :+1:

Reply:
@tschanzt As we've seen together, we need a relation to `potential_qty` in the `api.depends` of the compute function to ensure it's calculated correctly as recursive. So I suppose we still need to compute the components of each product to use this relation and ensure the components are computed in the right order.
IMO would be great to add a second test function based on `test_potential_qty__list` but with the creation of the components and finished product in the reverse order, to ensure we don't get the right result by a random order in the test.